src/Middleware/OutputCaching/src/OutputCacheEntryFormatter.cs

```diff
@@ -78,7 +79,21 @@ public static async ValueTask StoreAsync(string key, OutputCacheEntry value, Tim
 
         Serialize(bufferStream, formatterEntry);
 
-        await store.SetAsync(key, bufferStream.ToArray(), value.Tags ?? Array.Empty<string>(), duration, cancellationToken);
+        if (!bufferStream.TryGetBuffer(out var segment))
+        {
+            segment = bufferStream.ToArray();
+        }
+
+        var payload = new ReadOnlySequence<byte>(segment.Array!, segment.Offset, segment.Count);
+        if (store is IOutputCacheBufferStore bufferStore)
+        {
+            await bufferStore.SetAsync(key, payload, value.Tags, duration, cancellationToken);
+        }
+        else
+        {
+            // legacy API/in-proc: create an isolated right-sized byte[] for the payload
+            await store.SetAsync(key, payload.ToArray(), value.Tags, duration, cancellationToken);
```

Comment:
the API has always advertised nullability, and we only expect a single implementation, which already handles nullability; because of that, it seems redundant to force the null check; I don't mind adding it back, though